### PR TITLE
Add handling of tuples in type subscriptions

### LINKED ISSUE #162: Sphinx cross-references not substituted at deeper levels of nesting for types with nested iterables
I'm looking to use `sphinx-autodoc-typehints` with `nptyping`, and it seems that for types with nested iterables (which will be every use of the `NDArray` type), while the types are correctly resolved out of the annotation syntax at all levels of nesting, the Sphinx cross-reference substitutions are only made at the outermost level of nesting:

For example, with a signature and docstring of:

```
def basis_func(cls, eta: NDArray[(Any, ...), Float], *, n: int) -> NDArray[(Any, ...), Float]:
    """Generate the basis function over `eta` at the given `n`.

    :param eta: Array (multi-dimensional ok) of spatial position values
        over which to calculate basis function values

    :param n: Index value to use in basis function calculation

    """
```

The generated output is:

image

----

This is with Python 3.9.1 and:

```
Sphinx==3.4.3
sphinx-autodoc-typehints==1.11.1
nptyping==1.4.0
```

### PR DESCRIPTION
**Pull request by bskinn**
Fixes #162.

Initial implementation and tests added, and project lints fixed.

There are some significant questions to be resolved before merging. Prime among them is whether introducing `nptyping` as a test/dev dependency is a good idea.

Also, input on whether this implementation is sufficiently general, and whether additional/different test cases should be added/used would be good.

### CODE REVIEW

File: src/sphinx_autodoc_typehints/__init__.py [hunk 1]
```diff
@@ -150,7 +169,10 @@ def format_annotation(annotation: Any, config: Config) -> str:
         formatted_args = f"\\[{', '.join(repr(arg) for arg in args)}]"
 
     if args and not formatted_args:
-        fmt = [format_annotation(arg, config) for arg in args]
+        try:
+            fmt = [format_annotation(arg, config) for arg in args]
+        except TypeError:
```
Comment (bskinn) [inline src/sphinx_autodoc_typehints/__init__.py:174]: Also not entirely sure if this exception handling needs to be refined.

File: src/sphinx_autodoc_typehints/__init__.py [hunk 2]
```diff
@@ -89,6 +89,22 @@ def get_annotation_args(annotation: Any, module: str, class_name: str) -> tuple[
     return getattr(annotation, "__args__", ())
 
 
+def format_internal_tuple(t: tuple, config: Config) -> str:
+    # An annotation can be a tuple, e.g., for nptyping:
+    #   NDArray[(typing.Any, ...), Float]
+    # In this case, format_annotation receives:
+    #   (typing.Any, Ellipsis)
+    # This solution should hopefully be general for *any* type that allows
+    # tuples in annotations
```
Comment (gaborbernat) [inline src/sphinx_autodoc_typehints/__init__.py:98]: Please align it alongside 120 characters

File: src/sphinx_autodoc_typehints/__init__.py (hunk 1 above)
Comment (gaborbernat) [reply src/sphinx_autodoc_typehints/__init__.py:174]: Why do we need this?

**Comment by gaborbernat** [comment]
> Prime among them is whether introducing nptyping as a test/dev dependency is a good idea.

I see it's a pure wheel so I'm ok with this.

> Also, input on whether this implementation is sufficiently general, and whether additional/different test cases should be added/used would be good.

Can be a good starting place. We can add additional tests if users report bugs, no?

File: src/sphinx_autodoc_typehints/__init__.py (hunk 1 above)
Comment (bskinn) [reply src/sphinx_autodoc_typehints/__init__.py:174]: To handle things with bare integers in annotations, like `Float[32]` in `nptyping` and (presumably) [`Literal[640]` in PEP 646](https://www.python.org/dev/peps/pep-0646/#use-case-1-specifying-shape-values).

The `TypeError` here is raised because `args` is not iterable in these cases. Singletons are not an issue in other subscriptable types because they are special-cased in their handling earlier in the function.

I did the error trapping this way initially b/c EAFP model. I can think of two other ways to handle offhand at this point in the code, both of which are LBYL, but more precise:

1. Switch to a `try: iter(args); except ...` handler. Would make it more specifically catch the iterable/non-iterable difference.

2. Switch to an explicit `isinstance(args, Number)` check. This would be a very focused check, and would not cover cases with non-numeric singletons. I believe neither `nptyping` nor PEP 646 uses non-numeric singletons, but I'm not 100% sure; and, who knows what's out there in the ecosystem, or will be.

I think I'm inclined to revise to (1).

File: src/sphinx_autodoc_typehints/__init__.py (hunk 2 above)
Comment (bskinn) [reply src/sphinx_autodoc_typehints/__init__.py:98]: I'm not sure what you mean -- do you want me to remove the indentation?

**Comment by bskinn** [comment]
> > Prime among them is whether introducing nptyping as a test/dev dependency is a good idea.
> 
> I see it's a pure wheel so I'm ok with this.
> 

The main thing I was thinking of is, currently my test cases depend on some internal type transformations that `nptyping` does. If `nptyping` changes these transformations, that will cause this project's test suite to fail. I realized: I can just change the test examples so that `nptyping` does not transform them. That should remove the brittleness, at least most of the way. I'll make this change.

> > Also, input on whether this implementation is sufficiently general, and whether additional/different test cases should be added/used would be good.
> 
> Can be a good starting place. We can add additional tests if users report bugs, no?

Yep. That's what I was thinking.

File: src/sphinx_autodoc_typehints/__init__.py (hunk 1 above)
Comment (gaborbernat) [reply src/sphinx_autodoc_typehints/__init__.py:174]: Sounds good for 1 👍

File: src/sphinx_autodoc_typehints/__init__.py (hunk 2 above)
Comment (gaborbernat) [reply src/sphinx_autodoc_typehints/__init__.py:98]: yes; this line could reach all the way to 120 character, so no need to break it so soon

**Comment by gaborbernat** [comment]
> That should remove the brittleness, at least most of the way. I'll make this change.

Sounds good on this.

**Comment by bskinn** [comment]
For my reference:

- [ ] De-fold comment
- [ ] Re-work iterable/non-iterable check
- [ ] Fix typing problem
- [ ] Rewrite tests to avoid `nptyping` brittleness

File: src/sphinx_autodoc_typehints/__init__.py (hunk 2 above)
Comment (bskinn) [reply src/sphinx_autodoc_typehints/__init__.py]: Fixed in aa53976

**Comment by bskinn** [comment]
b279611 removes as much of the brittleness from the `nptyping` dependency as is possible with the current [`assert result == expected_result` check](https://github.com/tox-dev/sphinx-autodoc-typehints/blob/7b8c357be15d7b7b4ec4ad272f554f7ab3e0e197/tests/test_sphinx_autodoc_typehints.py#L209).

If `nptyping` restructures/refactors their internals, these test cases will break.

It would be possible to reduce/eliminate this risk of breakage by loosening the assert, perhaps only for the `nptyping` cases; e.g.,:

```

if "nptyping" in result:
    # Looser check that matches e.g. re.search(expected_regex, result)
    # where expected_regex has things in it like r"~nptyping\..+\.NDArray"
else:
    assert result == expected_result
```

That would make the test function messy, though.

I guess another alternative would be to make a new test function for the nptyping cases, with the above selective-matching behavior.

-----

TLDR; are you okay with accepting this possibility for breakage from `nptyping` for now, and we'll see what happens? Or, would you prefer to try to mitigate it now?

**Review by gaborbernat — APPROVED** [review]

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Please add a changelog with 1.16.0

https://github.com/tox-dev/sphinx-autodoc-typehints/blob/main/CHANGELOG.md

**Comment by bskinn** [comment]
> Please add a changelog with 1.16.0
> 
> https://github.com/tox-dev/sphinx-autodoc-typehints/blob/main/CHANGELOG.md

Done!

**Review by gaborbernat — APPROVED** [review]
